File: crates/photo-tauri/src/lib/stacks.ts
// 显示堆叠纯逻辑：将筛选+排序后的成员下标（filteredIndices）分组为网格显示项。
// 扫描模型不变（每文件一个 Capture，见 CONTEXT.md「画面」），堆叠是前端显示层分组。
// 两种堆叠模式（photo-config StackMode，配置于设置弹窗）：
// - ByFileName：同 stem（baseName）文件合并——JPG/NEF 同画面，key = baseName
// - ByTime：同组照片堆叠——拍摄时间差 ≤2s 的连拍合并（与 burst.ts 连拍阈值一致），
//   key = `t-<组内最早时间戳>`（比组序号稳定）
// - None：不堆叠，每成员独立成组（网格渲染路径统一，单成员组无堆叠 UI）
// 组语义：组位置 = 组内成员在显示序中的最小位置；激活成员默认 = 主格式（JPEG 优先，
// 出图格式优先于 RAW/其他）；用户手动切换后由 filter store 的 stackActive 覆盖。
import type { CaptureMeta } from './bindings'
import { parseExifDate } from './burst'

/** 堆叠组：网格中的一个显示项，成员 = 同组的 items 下标 */
export interface StackGroup {
  /** 分组键：ByFileName = baseName；ByTime = `t-<最早时间戳>`；None = `i-<下标>` */
  key: string
  /** 成员下标（captures.items 下标，按分组算法排序；size≥1） */
  members: number[]
  /** 激活成员下标（默认主格式；用户切换后由 filter store 的 stackActive 覆盖） */
  active: number
}

/** 主格式优先级：JPEG 出图优先；未命中列表（RAW/其他）rank=Infinity 排最后 */
const PRIMARY_ORDER = [
  'jpg',
  'jpeg',
  'png',
  'tif',
  'tiff',
  'gif',
  'webp',
  'bmp',
  'heic',
  'heif',
  'avif',
]

/** 同组照片堆叠的时间窗（ms）：连拍判定阈值，与 burst.ts 连拍分组一致（≤2s） */
export const STACK_TIME_GAP_MS = 2000

function extOf(path: string): string {
  const m = /\.([^.\\/]+)$/.exec(path)
  return m ? m[1].toLowerCase() : ''
}

/**
 * 选主成员：按主路径**真实扩展名**（非 primaryFormat 规范化名，避免 jpeg→.jpg
 * 假后缀，对齐 CaptureMeta::display_name 的取法）在 PRIMARY_ORDER 中取最优；
 * 全部未命中（如纯 RAW 堆叠）回退组内首个成员。
 */
function pickPrimary(members: number[], items: CaptureMeta[]): number {
  let best = members[0]
  let bestRank = Number.POSITIVE_INFINITY
  for (const i of members) {
    const rank = PRIMARY_ORDER.indexOf(extOf(items[i]?.primaryPath ?? ''))
    if (rank >= 0 && rank < bestRank) {
      bestRank = rank
      best = i
    }
  }
  return best
}

/** 组装 StackGroup 并记录组位置（组内成员在显示序中的最小位置，用于排序输出） */
function buildGroup(
  key: string,
  members: number[],
  items: CaptureMeta[],
  posOf: Map<number, number>,
): { g: StackGroup; pos: number } {
  let pos = Number.POSITIVE_INFINITY
  for (const i of members) {
    const p = posOf.get(i)
    if (p !== undefined && p < pos) pos = p
  }
  return { g: { key, members, active: pickPrimary(members, items) }, pos }
}

/**
 * 不堆叠：每成员独立成组（网格渲染路径与堆叠模式统一，单成员组无堆叠 UI）。
 * 空输入返回空数组。
 */
export function groupSingles(indices: number[]): StackGroup[] {
  return indices.map((i) => ({ key: `i-${i}`, members: [i], active: i }))
}

/**
 * 同文件名堆叠：显示序下标按 baseName 分组（组位置 = 组内首个成员位置，
 * 同 stem 因筛选/排序被打散时聚合到最先出现的位置）。
 * 单成员组也返回；无 baseName 的项跳过（防御，正常扫描不产生）。
 */
export function groupStacks(indices: number[], items: CaptureMeta[]): StackGroup[] {
  const byStem = new Map<string, number[]>()
  const stems: string[] = []
  for (const i of indices) {
    const stem = items[i]?.baseName ?? ''
    if (stem === '') continue
    const list = byStem.get(stem)
    if (list) {
      list.push(i)
    } else {
      byStem.set(stem, [i])
      stems.push(stem)
    }
  }
  const posOf = new Map<number, number>()
  indices.forEach((i, p) => posOf.set(i, p))
  const groups: { g: StackGroup; pos: number }[] = []
  for (const stem of stems) {
    groups.push(buildGroup(stem, byStem.get(stem)!, items, posOf))
  }
  groups.sort((a, b) => a.pos - b.pos)
  return groups.map((x) => x.g)
}

/**
 * 同组照片堆叠：按拍摄时间聚类（与显示序无关），相邻时间戳差 ≤ gapMs 归为同组。
 * - 组内成员按时间序（同时间戳保持显示序）；组 id = `t-<组内最早时间戳>`，
 *   比组序号稳定：筛选变化时最早帧仍在组内则 id 不变，覆盖激活不失效
 * - 组位置 = 组内成员在显示序中的最小位置（组按此排序输出）
 * - 无 dateTaken/解析失败的成员 = 独立单成员组（`x-<下标>`，不并入任何时间组）
 * 空输入返回空数组。
 */
export function groupByTime(
  indices: number[],
  items: CaptureMeta[],
  gapMs = STACK_TIME_GAP_MS,
): StackGroup[] {
  const withTime: { idx: number; t: number }[] = []
  const noTime: number[] = []
  for (const i of indices) {
    const t = parseExifDate(items[i]?.dateTaken ?? null)
    if (t === null) noTime.push(i)
    else withTime.push({ idx: i, t })
  }
  // 稳定按时间排序（同时间戳保持显示序；Array.sort 稳定）
  withTime.sort((a, b) => a.t - b.t)
  // 聚类：相邻差 ≤ gapMs 同组（与 burst.ts 连拍判定一致）
  const clusters: { members: number[]; t0: number }[] = []
  let cur: number[] = []
  let t0 = 0
  let prevT: number | null = null
  for (const e of withTime) {
    if (cur.length > 0 && prevT !== null && e.t - prevT > gapMs) {
      clusters.push({ members: cur, t0 })
      cur = []
    }
    if (cur.length === 0) t0 = e.t
    cur.push(e.idx)
    prevT = e.t
  }
  if (cur.length > 0) clusters.push({ members: cur, t0 })

  const posOf = new Map<number, number>()
  indices.forEach((i, p) => posOf.set(i, p))
  const groups: { g: StackGroup; pos: number }[] = []
  for (const c of clusters) {
    groups.push(buildGroup(`t-${c.t0}`, c.members, items, posOf))
  }
  for (const i of noTime) {
    groups.push(buildGroup(`x-${i}`, [i], items, posOf))
  }
  groups.sort((a, b) => a.pos - b.pos)
  return groups.map((x) => x.g)
}

File: crates/photo-ui/src/model/stacks.rs
//! 堆叠分组纯逻辑（对应 §4.3）。

use photo_domain::CaptureMeta;
use std::collections::{HashMap, HashSet};

use super::burst::parse_exif_date;

/// 堆叠组：网格中的一个显示项
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StackGroup {
    /// 分组键：ByFileName = baseName 或 baseName@dir；ByTime = `t-<最早时间戳>`；None = `i-<下标>`
    pub key: String,
    /// 成员下标（captures.items 下标，size >= 1）
    pub members: Vec<usize>,
    /// 激活成员下标（默认主格式优先，可手动覆盖）
    pub active: usize,
}

/// 主格式优先级：JPEG 出图优先；RAW/其他排最后
pub const PRIMARY_ORDER: &[&str] = &[
    "jpg", "jpeg", "png", "tif", "tiff", "gif", "webp", "bmp", "heic", "heif", "avif",
];

/// 连拍/同组照片堆叠时间窗（毫秒）
pub const STACK_TIME_GAP_MS: i64 = 2000;

fn ext_of(path: &str) -> String {
    if let Some(pos) = path.rfind('.') {
        let ext = &path[pos + 1..];
        if !ext.contains('/') && !ext.contains('\\') {
            return ext.to_ascii_lowercase();
        }
    }
    String::new()
}

fn dir_of(path: &str) -> &str {
    if let Some(pos) = path.rfind(|c| c == '/' || c == '\\') {
        &path[..pos]
    } else {
        ""
    }
}

/// 选主成员：按主路径真实扩展名在 PRIMARY_ORDER 中取最优；全部未命中回退组内首个成员
pub fn pick_primary(members: &[usize], items: &[CaptureMeta]) -> usize {
    if members.is_empty() {
        return 0;
    }
    let mut best = members[0];
    let mut best_rank = usize::MAX;

    for &i in members {
        if let Some(item) = items.get(i) {
            let ext = ext_of(&item.primary_path);
            if let Some(rank) = PRIMARY_ORDER.iter().position(|&x| x == ext) {
                if rank < best_rank {
                    best_rank = rank;
                    best = i;
                }
            }
        }
    }

    best
}

/// 不堆叠：每成员独立成组
pub fn group_singles(indices: &[usize]) -> Vec<StackGroup> {
    indices
        .iter()
        .map(|&i| StackGroup {
            key: format!("i-{i}"),
            members: vec![i],
            active: i,
        })
        .collect()
}

/// 同文件名堆叠：显示序下标按 (父目录, baseName) 分组
pub fn group_stacks(indices: &[usize], items: &[CaptureMeta]) -> Vec<StackGroup> {
    struct Bucket {
        dir: String,
        stem: String,
        members: Vec<usize>,
    }

    let mut buckets: HashMap<String, Bucket> = HashMap::new();
    let mut order: Vec<String> = Vec::new();
    let mut dirs_per_stem: HashMap<String, HashSet<String>> = HashMap::new();

    for &i in indices {
        let Some(item) = items.get(i) else { continue };
        let stem = item.base_name.clone();
        if stem.is_empty() {
            continue;
        }
        let dir = dir_of(&item.primary_path).to_string();
        let bucket_key = format!("{dir}\0{stem}");

        if let Some(bucket) = buckets.get_mut(&bucket_key) {
            bucket.members.push(i);
        } else {
            buckets.insert(
                bucket_key.clone(),
                Bucket {
                    dir: dir.clone(),
                    stem: stem.clone(),
                    members: vec![i],
                },
            );
            order.push(bucket_key);
        }

        dirs_per_stem.entry(stem).or_default().insert(dir);
    }

    let pos_of: HashMap<usize, usize> = indices.iter().enumerate().map(|(p, &i)| (i, p)).collect();
    let mut groups: Vec<(StackGroup, usize)> = Vec::new();

    for bucket_key in order {
        let b = buckets.remove(&bucket_key).unwrap();
        let key = if dirs_per_stem.get(&b.stem).map_or(0, |s| s.len()) > 1 {
            format!("{}@{}", b.stem, b.dir)
        } else {
            b.stem
        };

        let mut pos = usize::MAX;
        for &m in &b.members {
            if let Some(&p) = pos_of.get(&m) {
                if p < pos {
                    pos = p;
                }
            }
        }

        let active = pick_primary(&b.members, items);
        groups.push((
            StackGroup {
                key,
                members: b.members,
                active,
            },
            pos,
        ));
    }

    groups.sort_by_key(|(_, pos)| *pos);
    groups.into_iter().map(|(g, _)| g).collect()
}

/// 同组照片堆叠：按拍摄时间聚类，相邻时间戳差 <= gap_ms 归为同组
pub fn group_by_time(indices: &[usize], items: &[CaptureMeta], gap_ms: i64) -> Vec<StackGroup> {
    struct TimedItem {
        idx: usize,
        t: i64,
    }

    let mut with_time: Vec<TimedItem> = Vec::new();
    let mut no_time: Vec<usize> = Vec::new();

    for &i in indices {
        let date_taken = items.get(i).and_then(|meta| meta.date_taken.as_deref());
        match parse_exif_date(date_taken) {
            Some(t) => with_time.push(TimedItem { idx: i, t }),
            None => no_time.push(i),
        }
    }

    // 稳定按时间排序（同时间戳保持显示序）
    with_time.sort_by_key(|item| item.t);

    struct Cluster {
        members: Vec<usize>,
        t0: i64,
    }

    let mut clusters: Vec<Cluster> = Vec::new();
    let mut cur: Vec<usize> = Vec::new();
    let mut t0 = 0i64;
    let mut prev_t: Option<i64> = None;

    for item in with_time {
        if !cur.is_empty() && prev_t.is_some_and(|pt| item.t - pt > gap_ms) {
            clusters.push(Cluster {
                members: std::mem::take(&mut cur),
                t0,
            });
        }
        if cur.is_empty() {
            t0 = item.t;
        }
        cur.push(item.idx);
        prev_t = Some(item.t);
    }
    if !cur.is_empty() {
        clusters.push(Cluster { members: cur, t0 });
    }

    let pos_of: HashMap<usize, usize> = indices.iter().enumerate().map(|(p, &i)| (i, p)).collect();
    let mut groups: Vec<(StackGroup, usize)> = Vec::new();

    for c in clusters {
        let mut pos = usize::MAX;
        for &m in &c.members {
            if let Some(&p) = pos_of.get(&m) {
                if p < pos {
                    pos = p;
                }
            }
        }
        let active = pick_primary(&c.members, items);
        groups.push((
            StackGroup {
                key: format!("t-{t0}", t0 = c.t0),
                members: c.members,
                active,
            },
            pos,
        ));
    }

    for i in no_time {
        let pos = pos_of.get(&i).copied().unwrap_or(usize::MAX);
        groups.push((
            StackGroup {
                key: format!("x-{i}"),
                members: vec![i],
                active: i,
            },
            pos,
        ));
    }

    groups.sort_by_key(|(_, pos)| *pos);
    groups.into_iter().map(|(g, _)| g).collect()
}
